packages/react-app/src/views/QuadraticDiplomacyVote.jsx

```diff
@@ -1,176 +1,234 @@
-  const [currentPage, setCurrentPage] = useState(1);
+  const [currentStep, setCurrentStep] = useState(0);
   const [availableVoteTokens, setAvailableVoteTokens] = useState(0);
 
-  useEffect(async () => {
-    const initialContributors = contributorEntries.reduce((entries, current) => {
-      entries[current.wallet] = {
-        name: current.name,
-        voteTokens: 0,
-      }
-
-      return entries;
-    }, {})
-
-    setContributors(initialContributors);
-  }, [contributorEntries]);
+  // memoized values
+  const contributors = useMemo(
+    () =>
+      contributorEntries.reduce((entries, current) => {
+        entries[current.wallet] = {
+          name: current.name,
+          voteTokens: 0,
+        };
+        return entries;
+      }, {}),
+    [contributorEntries],
+  );
+  const allcontributorsSelected = useMemo(
+    () => Object.keys(contributors).length === Object.keys(selectedContributors).length,
+    [contributors, selectedContributors],
+  );
 
-  useEffect(async () => {
+  // side effects
+  useEffect(() => {
     if (voteCredits) {
       setAvailableVoteTokens(voteCredits.toNumber());
     }
   }, [voteCredits]);
 
-  const handleContributorSelection = (e, contributorAddress) => {
-    setSelectedContributors((prevSelectedContributors) => {
-      if (selectedContributors[contributorAddress]) {
-        const state = { ...prevSelectedContributors };
-        delete state[contributorAddress];
-        return state;
-      } else {
-        return {
-          ...prevSelectedContributors,
-          [contributorAddress]: contributors[contributorAddress]
+  // memoized callbacks
+  const handleSelectAllContributors = useCallback(
+    () => (allcontributorsSelected ? setSelectedContributors({}) : setSelectedContributors(contributors)),
+    [allcontributorsSelected, contributors, setSelectedContributors],
+  );
+  const handleContributorSelection = useCallback(
+    (e, contributorAddress) => {
+      setSelectedContributors(prevSelectedContributors => {
+        if (prevSelectedContributors[contributorAddress]) {
+          const state = { ...prevSelectedContributors };
+          delete state[contributorAddress];
+          return state;
+        } else {
+          return {
+            ...prevSelectedContributors,
+            [contributorAddress]: contributors[contributorAddress],
+          };
         }
-      }
-    })
-  }
-
-  const handleContributorVote = (e, op, clickedContributorAddress) => {
-    setAvailableVoteTokens((prevVotes) => {
-      return op === 'add' ? prevVotes - 1 : prevVotes + 1;
-    })
-
-    const newContributorVote = op === 'add'
-      ? selectedContributors[clickedContributorAddress].voteTokens + 1
-      : selectedContributors[clickedContributorAddress].voteTokens - 1;
-
-    setSelectedContributors((prevSelectedContributors) => {
-      return {
+      });
+    },
+    [contributors, setSelectedContributors],
+  );
+  const handleContributorVote = useCallback(
```

Comment:
@carletex: 
Do you think using `useCallback` here are redundant as well ? Given that the `handleSelectAllContributors` and `handleSubmitVotes` are passed directly to the `onChange` or `onClick` of component tree, I am not sure.

Reply:
I'd remove them as well.

The child components that use those functions (checkbox, button) are light, and I don't think re-rendering them will create performance issues.

> Don’t forget that useCallback() hook is called every time the component renders. Even useCallback() returning the same function object, still, the inline function is re-created on every re-rendering (useCallback() just skips it).

From: https://dmitripavlutin.com/dont-overuse-react-usecallback/